include/kodlab_mjbots_sdk/pid_controller.h

```diff
@@ -0,0 +1,533 @@
+  VectorNS get_dstate_setpoint(){return derivative_setpoint_;}
+
+  std::pair<VectorNS, VectorNS> get_setpoints(){
+    
+    return {setpoint_, derivative_setpoint_};
+  }
+
+  
+
+  /**
+   * @brief Function to update setpoints and vector PID output.
+   * @param state Current state for the controller.
+   * @param derivative_state Current derivative state of the controller.
+   * @param setpoint Desired state of the controller
+   * @param derivative_setpoint Desired derivative state of the controller
+   * @param time_step   Time difference between 2 updates
+   * @return Returns the Pid output
+   */
+
+  void Update(const VectorNS & state,const VectorNS & derivative_state,const VectorNS & setpoint,const VectorNS & 
+      derivative_setpoint,double time_step) {
+    
+    //Calling the setter function 
+    set_setpoint(setpoint,derivative_setpoint);
+    
+    // Calling Nominal Update function.
+    Update(state,derivative_state,time_step);
+  }
+
+  /**
+   * @brief Function to update setpoints and scalar PID output.
+   * @param state Current state for the controller.
+   * @param derivative_state Current derivative state of the controller.
+   * @param setpoint Desired state of the controller
+   * @param derivative_setpoint Desired derivative state of the controller
+   * @param time_step   Time difference between 2 updates
+   * @return Returns the Pid output
+   */
+  void Update(Scalar state,Scalar derivative_state,Scalar setpoint,Scalar derivative_setpoint,double time_step) {
+    //checking if using for scalar PID
+    static_assert(N==1, "Using Scalar function in vector PID");
+    set_setpoint(setpoint,derivative_setpoint);
+    
+    // Calling Nominal Update function.
+    Update(state,derivative_state,time_step);
+  }
+
+  /**
+   * @brief Function to update setpoints and PID output.
+   * @param state Current state for the controller.
+   * @param derivative_state Current derivative state of the controller.
+   * @param time_step   Time difference between 2 updates
+   * @return Returns the Pid output
+   */
+  virtual VectorNS Update(const VectorNS & state,const VectorNS & derivative_state,double time_step) {
+    // Calculate the errors
+    // Proportional error
+    error_ = setpoint_ - state;
+    
+    // Derivative error
+    d_error_ =  derivative_setpoint_ - derivative_state;
+    
+    
+    // Calling UpdateWithError.
+    return UpdateWithError(error_, d_error_,time_step);;
+  }
+
+  /**
+   * @brief Function to update setpoints and PID output.
+   * @param state Scalar state for the controller.
+   * @param derivative_state Scalar derivative state of the controller.
+   * @return Returns the Pid output
+   */
+  VectorNS Update(Scalar state,Scalar derivative_state) {
+    //checking if using for scalar PID
+    static_assert(N==1, "Using Scalar function in vector PID");
+    // converts scalar to vector form to calculate output
+    return Update(state*VectorNS::Ones(),derivative_state*VectorNS::Ones());
```

Comment:
Should have a time_step input I think ? 
And output a scalar
```suggestion
  Scalar Update(Scalar state, Scalar derivative_state) {
    //checking if using for scalar PID
    static_assert(N==1, "Using Scalar function in vector PID");
    // converts scalar to vector form to calculate output
    return Update(state*VectorNS::Ones(),derivative_state*VectorNS::Ones())(0);
```